lib/gbnp.gi

```diff
@@ -1,277 +1,170 @@
-
-      # Convert results back to path algebra
-      #  elements:
-      pgrob :=  QPA_Cohen2Path(grob,pa);
-
-      # Remove any zeroes we may've gathered from adding relations:
-      pgrob := Filtered(pgrob, x -> x <> Zero(pa));
-
-    else
-      Print("\nPlease make sure all elements are in the path algebra ",pa,
-            ".\n");
-      pgrob := false;
-    fi;
-
-    return pgrob;
+      i := i-1;
+    od;
 
+    return res;
   end
 );
 
 
-# Convert to format Cohen expects.
-InstallMethod( QPA_Path2Cohen,
-  "convert path algebra elements to Cohen reps",
-  true,
-  [ IsList ],
-  0,
-  function( els )
-    local e,i,oldrep,coefs,mons,creps;
-
-    creps := [];
-
-    for e in els do
-      coefs := [];
-      mons := [];
+InstallMethod( TipReducedList,
+  "for a list of path-algebra elements",
+  [ IsList, IsPathAlgebra ],
+  function(els, A)
+    local res, el, i;
 
-      oldrep := ExtRepOfObj(e)[2];
+    res := [];
 
-      for i in [ 2,4 .. Length(oldrep) ] do
-        Add(coefs,oldrep[i]);
-        Add(mons,oldrep[i-1]);
-      od;
+    for el in els do
+      if not IsZero(el) then
+        AddSet(res, el);
+      fi;
+    od;
 
-      Add(creps, [mons,coefs]);
+    i := Length(res);
+    while i > 0 do
+      if ForAny([1..i-1], j -> LeftmostOccurrence(TipWalk(res[i]), TipWalk(res[j])) <> fail) then
+        Remove(res, i);
+      fi;
+      i := i-1;
     od;
 
-    return creps;
+    return res;
   end
 );
 
 
-# Convert to format Cohen expects.
-#
-# NOTE: we are assuming that the elements in given
-#  list are from a path algebra isomorphic to a free
-#  algebra, i.e. there is only one vertex.
-#
-InstallMethod( QPA_Path2CohenFree,
-  "convert path (special case: free) algebra elements to Cohen reps",
-  true,
-  [ IsList ],
-  0,
-  function( els )
-    local e,i,oldrep,coefs,mons,creps;
+InstallMethod( RemainderOfDivision,
+  "for a path-algebra element and a list of path-algebra elements",
+  [ IsElementOfMagmaRingModuloRelations, IsList, IsPathAlgebra ],
+  function(y, X, A)
+    local r, n, y_tip, y_wtip, divided, i, p, u, v;
 
-    creps := [];
+    r := Zero(A);
+    n := Length(X);
 
-    for e in els do
-      coefs := [];
-      mons := [];
+    while not IsZero(y) do
+      y_tip := Tip(y);
+      y_wtip := TipWalk(y_tip);
 
-      oldrep := ExtRepOfObj(e)[2];
+      divided := false;
 
-      for i in [ 2,4 .. Length(oldrep) ] do
+      for i in [1..n] do
+        p := LeftmostOccurrence(y_wtip, TipWalk(X[i]));
 
-        Add(coefs,oldrep[i]);
+        if p <> fail then
+          u := Product(y_wtip{[1..p[1]-1]}, One(A));
+          v := Product(y_wtip{[p[2]+1..Length(y_wtip)]}, One(A));
 
-	# Check to see if this is our identity
-	# (same as our only vertex, our first
-	#  generator), if so correctly convert:
-	if ( oldrep[i-1] = [1] ) then
-          Add(mons,[]);
-	else
-          Add(mons,oldrep[i-1]);
-	fi;
+          y := y - TipCoefficient(y_tip)/TipCoefficient(X[i]) * u*X[i]*v;
```

Comment:
I don't know what data structures you guys use internally; but if this is a plain GAP `IsElementOfMagmaRingModuloRelations`, this part (head reduction) will be a major bottle neck in serious GB computations. There are ways to greatly improve that, though, by using appropriate data structures. 

Caveat: I worked on this in a rather different context (modules over group rings of polycyclic groups), and don't know much about QPA, but perhaps the following will be helpful for your anyway:

A relatively easy approach is to combine a hashtable with a heap (both available in optimized versions in the `datastructures` package): at the start of the reduction, one splits `y` into its "terms", and these into coefficient and "monomial". Then use the hashtable to map monomials to coefficients; and put the monomials of y in the heap, where the heap uses whatever termorder you want. This way, accessing the "tip" is trivial, you just get it from the heap; and you get the coefficient from the hashtable. The clever part then is how you perform reduction: you simply iterate over the terms `t` of (in this case) `X[i]` (assuming I guessed the meaning of the code here correctly; if I didn't, I hope the core of the idea still remains clear, though). Each term you multiply from left/right by `u`, `v`, etc; then look at the resulting term, split that into coefficient*monomial;  use that to update the coefficient for `monomial` stored in the hashmap (if it's not in there, add it; if it is in there, subtract from the existing coefficient); finally, push the monomial into the heap.

This avoids creating many costly objects, and more importantly, replaces the costly computation (possibly requiring linear time in the number of terms, depending on the your datastructure) of the tip of `y` by constant time; avoids assembling lots of temporary objects (which is super expensive in large computations), and basically shaves of a linear or quadratic factor in the runtime (depending on the existing data structures being used); while of course the original problem remains doubly exponential or worse (in the non-commutative setting you are in...), in practice these gains can make a difference between a problem taking days, or just minutes.

Reply:
Thank you for the detailed explanations!

I should have stressed that this is merely my personal workaround, which I propose here as a temporary fix for #20. I am not involved in the development of QPA.

This fix is certainly not a good implementation for serious GB computations, but in my experience QPA is not the right choice for such computations anyway (using GBNP directly is much faster).

Please feel free to close this. In particular, if someone who knows QPA has a better solution.

Reply:
Thank you very much for all your comments.

In the outset QPA was designed to work with whatever package computing Groebner bases. This you see in the command  GroebnerBasis( I, grb ), which is just telling QPA that the list elements in grb is a Groebner basis for the ideal I.  So then you can enter this algebra into QPA as:

Q := Quiver( 3,[ [ 1, 2, "a" ], [ 1, 2, "b" ], [ 2, 2, "c" ], [ 2, 3, "d" ], [ 3, 1, "e" ] ] );
KQ := PathAlgebra( Rationals, Q );
AssignGeneratorVariables(KQ);
relations := [ d*e, c^2, a*c*d - b*d, e*a ];
I := Ideal( KQ, relations );
# Use any package to compute a Groebner basis for I:
gb := AnyGroebnerBasis( relations, KQ );
# Tell QPA that this is a Groebner basis for I
gbb := GroebnerBasis( I, gb );
A := KQ/I;

In the second version of QPA, [QPA2](https://github.com/oysteins/QPA2), we have made our own implementation of a Groebner basis algorithm (with the same starting point as above). 
We would be very happy if you also could have a look at QPA2, to see if what you want to do is possible there and if not, what is missing.  QPA2 is under development, so it does not have all the functionality as QPA1, but we are striving to get there and beyond. 

Best regards, the QPA-team.

Reply:
As expected, the examples given in #20 work just fine in QPA2.

Maybe it would be helpful if QPA2 could also compute the reduced version of a Gröbner basis? Being able to compare paths (`kQ.a < kQ.b` doesn't work at the moment) would be a good feature, too.

I am not sure whether the performance of ComputeGroebnerBasis in QPA2 will be an issue. However, the remarks fingolfin made above should equally well apply there, since the implementation seems rather high level.